Fix reschedule so that it won't break nested locks

andyross opened the pull request:
First cut at a fix for #16273.  I managed to convince myself that the original suggested fix (check interrupt state always and only ever call swap when an irq_unlock of the current key would unmask interrupts) was correct in all cases, and that the original behavior can only be seen as a bug.  My concerns about compatible behavior live on only as an editorial comment in the commit message.

NOTE THAT THIS CAN'T MERGE YET.  The new z_arch_irq_unlocked() predicate isn't implemented yet on ARC, because I can't find docs and don't want to write the code blindly to simulator behavior (and, to be fair, don't want to go through the trouble of getting nsim working).  Someone expert (@ruuddw? @vonhust?) should really look at this.  As it ARC code will build and match the pre-existing (buggy!) behavior, but the new test case in kernel/common will fail by design.

Commit:
arch/*: Add z_arch_irq_unlocked() predicate and test

It's useful to be able to inspect the key returned from
z_arch_irq_unlock() to see if interrupts were enabled at the point
where z_arch_irq_lock() was called.  Architectures tend to represent
this is a simple way that doesn't require platform assembly to
inspect.

Adds a simple test to kernel/common that validates this predicate with
a nested lock.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Don't reschedule inside a nested lock

The internal "reschedule" API has always understood the idea that it
might run in a ISR context where it can't swap.  But it has always
been happy to swap away when in thread mode, even when the environment
contains an outer lock that would NOT be expecting to swap!  As it
happened, the way irq locks are implemented (they store flag state
that can be restored without context) this would "work" even though it
was completely breaking the synchronization promise made by the outer
lock.

But now, with spinlocks, the error gets detected (albeit in a clumsy
way) in debug builds.  The unexpected swap triggers SPIN_VALIDATE
failures in later threads (this gets reported as a "recursive" lock,
but what actually happened is that another thread got to run before
the lock was released and tried to grab the same lock).

Fix this so that swap can only be called in a situation where the irq
lock key it was passed would have the effect of unmasking interrupts.
Note that this is a real behavioral change that affects when swaps
occur: it's not impossible that there is code out there that actually
relies on this "lock breaking reschedule" for correct behavior.  But
our previous implementation was irredeemably broken and I don't know
how to address that.

Fixes #16273

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

carlescufi reviewed (commented):
Thanks for this PR @andyross!

I wonder if this would be a good time to introduce a `z_arch_irq_unlock_all()` at the same time, since this will be needed to fix https://github.com/zephyrproject-rtos/zephyr/issues/8393 (since we are adding irq-related code to all archs)

pizi-nordic commented on kernel/sched.c:
```diff
@@ -502,7 +502,7 @@ void z_thread_priority_set(struct k_thread *thread, int prio)
 }
 
-static inline int resched(void)
+static inline int resched(u32_t key)
```
const u32_t key ?

pdunaj replied:
this is not a pointer, does it matter really?

pizi-nordic replied:
This is function which tests some conditions. Since you testing variable, const is IMHO what we should have here. But as you said: this is not a pointer, so my comment is just suggestion.

pizi-nordic commented on arch/posix/include/posix_soc_if.h:
```diff
@@ -37,6 +37,11 @@ void posix_irq_full_unlock(void);
 int  posix_get_current_irq(void);
 /* irq_offload() from irq_offload.h must also be defined by the SOC or board */
 
+static inline bool z_arch_irq_unlocked(unsigned int key)
```
const unsigned int key?

andrewboie replied:
how is that helpful? it's passed by value

ruuddw commented on include/arch/arc/v2/irq.h:
```diff
@@ -127,6 +127,12 @@ static ALWAYS_INLINE void z_arch_irq_unlock(unsigned int key)
+static ALWAYS_INLINE bool z_arch_irq_unlocked(unsigned int key)
+{
+	/* FIXME: need docs on key format! */
+	return false;
```
key contains the relevant ARC status32 bits: interrupt enable (IE) and interrupt masking level.
The bit to test would be IE. This is bit 31 (0x8000_0000); 1=enabled; 0=disabled;

vonhust commented:
@andyross @ruuddw , for ARC , the code should be 
```
static ALWAYS_INLINE bool z_arch_irq_unlocked(unsigned int key) 
{ 
         /* ARC irq lock uses instruction "clri r0",  
          * r0 ==  {26’d0, 1’b1, STATUS32.IE, STATUS32.E[3:0] } 
          * bit4 is used to record IE (Interrupt Enable) bit
          */
         return (key & 0x10)  ==  0x10;
}
```

pizi-nordic commented on include/arch/x86/arch.h:
```diff
@@ -453,6 +453,15 @@ static ALWAYS_INLINE void z_arch_irq_unlock(unsigned int key)
 	z_do_irq_unlock();
 }
 
+/**
```
IMHO this comment should be before every z_arch_irq_unlocked() definition, as the function name might be misleading.
Have you considered different name? For example z_arch_irq_was_unlocked() ?

carlescufi commented:
@andyross could you take a look at the review comments?  This fixes a high priority bug, would be great if we could get it merged soon. Thanks!

nordic-krch commented:
Imo, this is half way solution as we still need `k_is_in_atomic()` for certain situations.
We need something like `z_arch_irq_is_locked(void)` which would read directly from the hardware. For the first round `z_arch_irq_is_locked(void)` can be implemented sub-optimally as:
```
bool k_is_in_atomic(void)
{
  if (k_is_in_isr()) return true;
  int k = irq_lock();
  irq_unlock(key);
  return (k != 0); //key value is probably arch dependent so we might need arch specific code here
}
```

pdunaj commented:
Hi @nordic-krch , `resched` is checking both `z_arch_irq_unlocked(key)` and `!z_is_in_isr()` so it seems correct. The is_atomic API may be nice to have as a convenience. The way it is proposed in the snippet would be problematic as it locks and unlocks the IRQs which takes time (2us or so if I remember correctly) and can be pain if something calls is_atomic frequently. Anyway what @andyross proposed should be fine.

pdunaj commented:
If there is no consensus how to deal with other architectures maybe we can leave them with
```
#define z_arch_irq_unlocked(x) false
```
and allow others to do the right fix.
Nothing will change for them while ARM will be fixed. Bug can be kept open with information which architectures are not fixed.

I personally believe that for most popular archs such critical issues should be fixed asap without waiting for others.

nordic-krch commented:
> API may be nice to have as a convenience

I don't think it's a convenience. We have that in #16194 which is currently blocked because lack of this feature. We don't know if we can wait or not. And i'm not saying that this lock/unlock pair should be used internally. I know this is slow. That's why i said for `for the first round ...`. But if there would be a problem with providing arch specific readout of current interrupt locking state then i would prefer to have functionality than have it delayed.

pdunaj commented:
@nordic-krch I guess you are right. Though with the proposed change this could be derived.

I think, it would move much faster if PR would be done for specific architectures only. I understand Zephyr is targeting many things but I see no reason why changes like this one target all archs should be done in one PR or by one person only.
I think this is the essential reason for this change not being in yet.

andyross commented:
New version pushed that includes the ARC implementation (@ruudw and @vonhust, if you could test this on your setup and report success, I'd be appreciative -- I don't have the ability to run ARC tests easily) and addresses most of the review comments.

Note that there are failures in some platforms in tests/kernel/fatal.  The proximate cause here is that exception handlers on those platforms appear to be returning false for z_is_in_isr(), causing a k_thread_abort() call in the test to try to context switch (!!! -- I think every one of these platforms is on the interrupt stack when this happens!).  But that sort of worked, because it was a thread abort of _current and nothing would ever try to return back into it.  But now we **don't** switch, and the code continues on to some other failure it never reached, resulting in a recursive panic.

Basically something bad is wrong with exception entry on some platforms, and it's blocking this patch.  Working on it now.

carlescufi commented:
>It's useful to be able to inspect the key returned from z_arch_irq_unlock() 

I think you mean `z_arch_irq_lock()`

carlescufi commented:
> Imo, this is half way solution as we still need `k_is_in_atomic()` for certain situations.
> We need something like `z_arch_irq_is_locked(void)` which would read directly from the hardware. For the first round `z_arch_irq_is_locked(void)` can be implemented sub-optimally as:
> 
> ```
> bool k_is_in_atomic(void)
> {
>   if (k_is_in_isr()) return true;
>   int k = irq_lock();
>   irq_unlock(key);
>   return (k != 0); //key value is probably arch dependent so we might need arch specific code here
> }
> ```

@andyross @nordic-krch 
Can't we have both? `z_arch_irq_unlocked(unsigned int key)` to find out whether interrupts were enabled before we called `irq_lock()` and then a new `k_irq_is_locked(void)` that is implemented with an arch-specific `z_arch_irq_is_locked(void)` that reads directly from the hardware?

andyross commented:
@nordic-krch: what's the situation you're thinking of?  Reading mask state from the hardware isn't cheap on every platform.  And design-wise, being in a situation where you need to know what the lock state is and don't have access to the key to unlock it tells me that you're in a critical section that is WAY too big anyway. :)

Using the key was chosen deliberately, basically.   It's faster and safer.

carlescufi commented:
@andyross I think @nordic-krch is trying to find a way to know when it is safe to wait on a kernel object. Right now in the proposal he uses `k_is_in_isr()` but this is not enough if it's thread mode but with interrupts disabled, since `LOG_` can be called from anywhere:

https://github.com/zephyrproject-rtos/zephyr/pull/16194/files#diff-b49ebd7b7d98a0d159b2a4cb8b974c75R329

nordic-krch commented:
@andyross, The case is as @carlescufi pointed out. What about having `k_irq_is_locked(unsigned int *key)` if key=NULL then hardware readout is performed using something like `z_arch_irq_is_locked(void)`.

andyross commented:
I'm not really opposed to the API, it's more of an aesthetic complaint.  I mean, on existing architectures (all of them) it's just going to be implemented with the equivalent of:

    bool k_irq_is_locked(void)
    {
        unsigned int key = z_arch_irq_lock();
        bool ret = z_arch_irq_unlocked(key);

        z_arch_irq_unlock(key);
        return ret;
    }

The only thing an optimized version would save is the resulting unlock call.  And on a few platforms it's possible to read the mask flags a little faster than it is to read-and-set them (e.g. on x86 it saves a CLI instruction).

But really, this is cheap (or at least not significantly more expensive than the custom API requested), so I guess I'm wondering if it's really worth having this as an API to be supported by every architecture.

Alternatively: surely you're doing your own synchronization in that logging code already somewhere, right?  You don't have a irq lock key (or a spinlock key -- same thing modulo the struct wrapper) you can use as a proxy?

carlescufi commented:
> I'm not really opposed to the API, it's more of an aesthetic complaint. I mean, on existing architectures (all of them) it's just going to be implemented with the equivalent of:
> 
> ```
> bool k_irq_is_locked(void)
> {
>     unsigned int key = z_arch_irq_lock();
>     bool ret = z_arch_irq_unlocked(key);
> 
>     z_arch_irq_unlock(key);
>     return ret;
> }
> ```
> 
> The only thing an optimized version would save is the resulting unlock call. And on a few platforms it's possible to read the mask flags a little faster than it is to read-and-set them (e.g. on x86 it saves a CLI instruction).
> 
> But really, this is cheap (or at least not significantly more expensive than the custom API requested), so I guess I'm wondering if it's really worth having this as an API to be supported by every architecture.
> 
> Alternatively: surely you're doing your own synchronization in that logging code already somewhere, right? You don't have a irq lock key (or a spinlock key -- same thing modulo the struct wrapper) you can use as a proxy?

I really don't want to block this PR because of this parallel discussion (I re-triggered CI but it seems there's still a couple of failures) but at least on Arm, checking whether interrupts are enabled is a single instruction `mrs key, {PRIMASK, BASEPRI}` so IMHO it would be substantially less code executed? Maybe I'm missing something, but why would we need to disable and then reenable interrupts just to check whether those are enabled at a particular point in time for archs that have a trivial mechanism to check for that?

andyross commented:
> Maybe I'm missing something, but why would we need to disable and then reenable interrupts just to check whether those are enabled 

Mostly because I'm 75% confident you already are somewhere in that logging code, for the obvious reason that the logging is going to need some kind of internal synchronization anyway.  My resistance is to adding a new API that every architecture needs to implement, the example there was just to show that this is an easy thing to work around.

FWIW: a really good example of why API minimalism and avoidance of per-platform hackery is important is precisely the remaining failures you're seeing on this PR.

What's happening in those tests is that user code is calling Z_ARCH_EXCEPT() to generate a synchronous exception, but on a bunch of architectures this call **does not**, in fact, generate an exception at all and just invokes z_NanoFatalErrorHandler() on the local stack and just ... assumes that user code will arrange so that it never returns (in this case by calling k_thread_abort(_current)).  But the z_swap() call in k_thread_abort() is now doing a reschedule that is aware of the lock state, and the new code is refusing to actually swap away because of an outer lock (that is in app code, not kernel code), so we "return" from Z_ARCH_EXCEPT() and things blow up.

That is: someone at some point decided to add an API because it was immediately useful, but didn't implement it the same way everywhere.  And now your super high priority bug is blocked because I need to back out and figure out a way to implement a more robust fallback hack than the one we had already (it's looking right now that it's going to have to be the equivalent of z_arch_irq_unlock_all(), hilariously -- either that or figure out how to implement a proper trap on all architectures that don't do it already).

carlescufi commented:
> > Maybe I'm missing something, but why would we need to disable and then reenable interrupts just to check whether those are enabled
> 
> Mostly because I'm 75% confident you already are somewhere in that logging code, for the obvious reason that the logging is going to need some kind of internal synchronization anyway. 

Not sure what you mean here with "you already are". You mean that the logging subsystem is already disabling interrupts? That is correct, as far as I can tell both the `LOG_` functions and the background thread do disable interrupts to synchronize on the list that holds the logs, so I think we could use that key as you say. @nordic-krch thoughts?

>My resistance is to adding a new API that every architecture needs to implement, the example there was just to show that this is an easy thing to work around.
[...]
> That is: someone at some point decided to add an API because it was immediately useful, but didn't implement it the same way everywhere. And now your super high priority bug is blocked because I need to back out and figure out a way to implement a more robust fallback hack than the one we had already (it's looking right now that it's going to have to be the equivalent of z_arch_irq_unlock_all(), hilariously -- either that or figure out how to implement a proper trap on all architectures that don't do it already).

Point taken, thanks for explaining.

Commit:
kernel/thread_abort: Swap, don't reschedule when aborting _current

The z_reschedule() call (as of the accompanying fix) will not swap
away from a thread if called with a nested irq lock held.

But for the specific case of aborting the current thread, we
absolutely need to swap regardless of how many locks the thread that
just aborted might have held.  So call z_swap() explicitly here.

This preserves the existing z_reschedule() call in other circumstances
for compatibility with existing test cases, but adds a note explaining
why it's there when the only obvious reason for it is already covered.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

carlescufi commented:
>it's looking right now that it's going to have to be the equivalent of z_arch_irq_unlock_all(), hilariously

FWIW you already had an initial (very simple and effective) implementation of this for Arm and x86 here: https://github.com/zephyrproject-rtos/zephyr/pull/9668/files

Commit:
tests/kernel/fatal: Fix wait-for-interrupt delay in stack check test

Contrary to the comment in code, this test is NOT, in fact, compiled
with a traditional ticked kernel.  Spinning won't work reliably
because interrupts won't necessarily be delivered when you expect.
This test case would fail spuriously as I moved things around when
debugging.

Doing it right (using a k_timer in this case) is actually less code
anyway.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/fatal: Clean up z_except_reason() fallback implementation

Architectures that lack implementations of synchronous traps (via
Z_ARCH_EXCEPT()) end up using a z_except_reason() implementation that
doesn't actually trap at all.  It just invokes
z_NanoFatalErrorHandler() in the current thread context.

That has two problems:

First, it was just blindly assuming that the error handling invoked
would abort the current thread, swap away, and never return.  But that
can be application code in z_SysFatalErrorHandler that we can't
control.

Second, it was too broad with this assumption and stuff a
CODE_UNREACHABLE hint in for the compiler.  But in fact
z_except_reason() may be invoked in interrupt context (for example the
stackprot check) where it may NOT swap away and WILL return
synchronously from the call.  This doesn't seem to have caused a
miscompilation in production code, but it made a total voodoo hash out
of my debugging around this macro for an hour or so until I figured
out why my logging was being optimized out.

Do the abort unconditionally instead of relying on the app, and remove
the incorrect compiler hint.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
OK, I think I have this working now.  Explanations for the issues fixed are in the separate commit messages.  Basically there were two spots where code (as I feared) really was expecting the old "reschedule can swap when doubly-locked" behavior.  But it was kernel internals without API exposure, so we can fix it internally.

nashif commented on tests/kernel/fatal/src/main.c:
add parens

andyross replied:
I swear I thought there was some exception for empty blocks? :)

Regardless, need to correct whitespace anyway (been mucking with my emacs config and apparently broke the C style autodetect stuff)

carlescufi commented:
@andyross thanks for following-up.

>Mostly because I'm 75% confident you already are somewhere in that logging code, for the obvious reason that the logging is going to need some kind of internal synchronization anyway.

I looked at the code again, and I think your proposal although reasonable, won't work in this case. This is because the log message needs to be allocated from a mem slab *before* you can actually put in in a list, which is where interrupts are being locked to protect that list from being accessed from two contexts. And it's really the call to `k_mem_slab_alloc()` that needs to be called either with a timeout or `K_NO_WAIT` depending on whether interrupts are enabled or not. So, at least in this particular case, I don't see an easy way to reuse the locking to then check if interrupts were enabled with the key you get from it, due to the order in which operations happen.

carlescufi reviewed (approved):
Solid fix, additional functionality can come in follow-up reviews.

nordic-krch commented:
@carlescufi the case is as you described.

nashif commented:
@andyross removed DNM here

ruuddw commented:
> New version pushed that includes the ARC implementation (@ruudw and @vonhust, if you could test this on your setup and report success, I'd be appreciative -- I don't have the ability to run ARC tests easily) and addresses most of the review comments.

I've tested and it works for me. Only a minor style consistency comment, see inline remark.

ruuddw commented on include/arch/arc/v2/irq.h:
```diff
@@ -127,6 +127,19 @@ static ALWAYS_INLINE void z_arch_irq_unlock(unsigned int key)
+	 * r0 ==  {26’d0, 1’b1, STATUS32.IE, STATUS32.E[3:0] }
+	 * bit4 is used to record IE (Interrupt Enable) bit
+	 */
+	return (key & 0x10)  ==  0x10;
```
Style of this check is different from the other architectures, others use (value & mask) !=0
Suggest to change this one (though the compiler migth generate similar optimized code from either).

nategraff-sifive commented on include/arch/riscv32/arch.h:
```diff
@@ -115,6 +115,22 @@ static ALWAYS_INLINE void z_arch_irq_unlock(unsigned int key)
+	 * the argument in z_arch_irq_unlock() that has me worried
+	 * that something elseswhere might try to set a bit?  Do it
+	 * the safe way for now.
+	 */
```
If I were to guess why the mask is applied in `z_arch_irq_unlock()`, it's not because something might set a bit in `key`, but because other fields of `mstatus` might change, and we only want to affect the `IEN` bit.
